pkg/kube/informerfactory/factory.go

```diff
@@ -0,0 +1,222 @@
+// * Ability to run a single informer rather than all of them.
+package informerfactory
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pkg/kube/kubetypes"
+	"istio.io/istio/pkg/util/sets"
+)
+
+// NewInformerFunc returns a SharedIndexInformer.
+type NewInformerFunc func() cache.SharedIndexInformer
+
+type StartableInformer struct {
+	Informer cache.SharedIndexInformer
+	start    func(stopCh <-chan struct{})
+}
+
+func (s StartableInformer) Start(stopCh <-chan struct{}) {
+	s.start(stopCh)
+}
+
+// InformerFactory provides access to a shared informer factory
+type InformerFactory interface {
+	// Start initializes all requested informers. They are handled in goroutines
+	// which run until the stop channel gets closed.
+	Start(stopCh <-chan struct{})
+
+	// InformerFor returns the SharedIndexInformer the provided type.
+	InformerFor(resource schema.GroupVersionResource, opts kubetypes.InformerOptions, newFunc NewInformerFunc) StartableInformer
+
+	// WaitForCacheSync blocks until all started informers' caches were synced
+	// or the stop channel gets closed.
+	WaitForCacheSync(stopCh <-chan struct{}) bool
+
+	// Shutdown marks a factory as shutting down. At that point no new
+	// informers can be started anymore and Start will return without
+	// doing anything.
+	//
+	// In addition, Shutdown blocks until all goroutines have terminated. For that
+	// to happen, the close channel(s) that they were started with must be closed,
+	// either before Shutdown gets called or while it is waiting.
+	//
+	// Shutdown may be called multiple times, even concurrently. All such calls will
+	// block until all goroutines have terminated.
+	Shutdown()
+}
+
+// NewSharedInformerFactory constructs a new instance of informerFactory for all namespaces.
+func NewSharedInformerFactory() InformerFactory {
+	return &informerFactory{
+		informers:        map[informerKey]cache.SharedIndexInformer{},
+		startedInformers: sets.New[informerKey](),
+	}
+}
+
+// InformerKey represents a unique informer
+type informerKey struct {
+	gvr           schema.GroupVersionResource
+	labelSelector string
+	fieldSelector string
+	informerType  kubetypes.InformerType
+	namespace     string
+}
+
+type informerFactory struct {
+	lock      sync.Mutex
+	informers map[informerKey]cache.SharedIndexInformer
+	// startedInformers is used for tracking which informers have been started.
+	// This allows Start() to be called multiple times safely.
+	startedInformers sets.Set[informerKey]
+
+	// wg tracks how many goroutines were started.
+	wg sync.WaitGroup
+	// shuttingDown is true when Shutdown has been called. It may still be running
+	// because it needs to wait for goroutines.
+	shuttingDown bool
+}
+
+var _ InformerFactory = &informerFactory{}
+
+func (f *informerFactory) InformerFor(resource schema.GroupVersionResource, opts kubetypes.InformerOptions, newFunc NewInformerFunc) StartableInformer {
```

Comment:
Return nil when it is shutting down

Reply:
The k8s one  doesn't. Not sure we want that - if we do, every code needs to handle the informer being nil. You should always be able to get an informer, just may not be able to start it

Reply:
It is resonable to return nil if the factory is shuttingdown. 

And you mean return a informer, just has no effect calling Start afterwards? That also works

Reply:
Yes, that's how this and the upstream factory works